File: triangle.cc
#include <vector>
using namespace std;

class Solution {
public:
  int minimumTotal(vector<vector<int> >& triangle) {
    vector<int> min_sum(triangle.size());
    for (int i = 0; i < triangle.size(); i++) {
      int prev = min_sum[0];
      min_sum[0] += triangle[i][0];
      for (int j = 1; j < i; j++) {
        int temp = min_sum[j];
        min_sum[j] = min(prev, min_sum[j]) + triangle[i][j];
        prev = temp;
      }
      min_sum[i] = prev + triangle[i][i];
    }
    int ans = INT_MAX;
    for (int i = 0; i < triangle.size(); i++) {
      if (min_sum[i] < ans) ans = min_sum[i];
    }
    return ans;
  }
};

File: triangle.java
public class Solution {
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        int[] minSum = new int[triangle.size()];
        minSum[0] = triangle.get(0).get(0);
        for (int i = 1; i < triangle.size(); i++) {
            ArrayList<Integer> row = triangle.get(i);
            minSum[i] = minSum[i - 1] + row.get(i);
            for (int j = i - 1; j > 0; j--)
                minSum[j] = Math.min(minSum[j], minSum[j - 1]) + row.get(j);
            minSum[0] += row.get(0);
        }
        int ans = minSum[0];
        for (int i = 1; i < minSum.length; i++)
            if (minSum[i] < ans) ans = minSum[i];
        return ans;
    }
}
